File: problem_sheet4/task2b.py
import sys
import matplotlib.pyplot as plt
import os
import logging
import pickle
import numpy as np
import numba
# from numba import jit, njit
# from numba.typed import List
from igraph import Graph

logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)

# create console handler and set level to debug
ch = logging.StreamHandler()
ch.setLevel(logging.INFO)
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
ch.setFormatter(formatter)
logger.addHandler(ch)

MCS_STEPS = 1000

# @jit(nopython=False)
def agent_move(node, p, q, states, neighbours):
    if np.random.random_sample() < p:
        if np.random.random_sample() < 0.5:
            states[node] = states[node]*(-1)
    else:
        if len(neighbours) < q:
            return
        voters = np.random.choice(neighbours, q, replace=False)
        S = 0
        for i in voters:
            S += states[i]
        if S == q:
            states[node] = 1
        elif S == -q:
            states[node] = -1

# @jit(nopython=False)
def mcs(adj_l, N, p, q, states):
    for _ in range(N):
        # choose random node
        node = np.random.randint(0, N)
        neigh = adj_l[node]
        agent_move(node, p, q, states, neigh)
    # return calc_conc(states, N)

# @jit(nopython=False)
def simulate(adj_l, states, N, p, q, mcs_steps):
    for _ in range(mcs_steps):
        mcs(adj_l, N, p, q, states)
    return calc_conc(states, N)

# @jit(nopython=False)
def gen_states(N, c):
    rang = np.array([x for x in range(N)])
    states = np.array([-1 for x in rang])
    for x in np.random.choice(rang, int(c * N), replace=False):
        states[x] = 1
    return states

# @jit(nopython=False)
def calc_conc(states, N):
    return (1/(2*N)) * states.sum() + (1/2)

def save_results(N, q, c, result, p=100):
    logger.info("Saving results...")
    result = [x for x in result]
    if not os.path.exists("results3"):
        os.mkdir("results3")
    if p == 100:
        with open(f"results3/{N}_{q}_{c}", "wb+") as f:
            pickle.dump(result, f)
    else:
        with open(f"results3/{N}_{q}_{c}_{p}", "wb+") as f:
            pickle.dump(result, f)

def pyth_wrapper(N, p, q, avg_d, c):
    # Create graph and needed adj_list
    g = Graph.Erdos_Renyi(N, avg_d/(N-1))
    logger.info(f"Generating adj list for N={N}, p={p}, init_c = {c}, q={q}...")
    adj_l = []
    for x in g.get_adjlist():
        adj_l.append(np.array(x))
    logger.info("Creting adj list is done!!")
    states = gen_states(N, c)
    logger.info("Starting simulation...")
    return simulate(adj_l, states, N, p, q, MCS_STEPS)
    

if __name__ == "__main__":
    logger.info("Strarting simulation.")
    if len(sys.argv) != 4:
        ## INPUT: N, q, c
        raise ValueError
    else:
        concentration = list()
        for p in np.linspace(0.0, 1.0, 20):
            res = pyth_wrapper(
                N=int(sys.argv[1]),
                p=p,
                q=int(sys.argv[2]),
                avg_d=14,
                c=float(sys.argv[3])
            )
            logger.info(f"Result = {res}")
            save_results(
                sys.argv[1],
                sys.argv[2],
                sys.argv[3],
                concentration,
                p=p
            )
            concentration.append(res)
        save_results(sys.argv[1], sys.argv[2], sys.argv[3], concentration)

File: problem_sheet4/julia/task2b.jl
# This file is very the same as the python script
# It only differs in syntax.

using Random
Random.seed!(2019); # make sure this tutorial is reproducible

using LightGraphs
using StatsBase

N = 100000
avg_d = 14

MCS_STEPS = 10000

c = parse(Float64, ARGS[1])
println("Con c(0)=$c")

q = parse(Int, ARGS[2])
println("Neighbours to consider q=$q")


function agent_move(
    node::Int64,
    p::Float64,
    q::Int64,
    states::Array,
    neighbours::Array
    )
    if Random.rand() < p
        # If p then with prob 1/2 = 1 and with 1/2 = -1
        if Random.rand() < 0.5
            states[node] = states[node] *(-1)
        end
    else
        # Check if we can select q neighbors
        if length(neighbours) < q
            return
        end
        # Get randomly q neighbours
        voters = sample(neighbours, q, replace=false)
        # Check if we should change the state
        S = 0
        for v in voters
            S += states[v]
        end
        if S == q
            states[node] = 1
        elseif S == -q
            states[node] = -1
        end
    end
end

function mcs(graph, N, p, q, states)
    for i in 1:N
        # picka a node
        node = rand(1:N)
        agent_move(
            node,
            p,
            q,
            states,
            neighbors(graph, node)
        )
    end
end

function simulate(graph, N, p, q, n_mcs, states)
    println("Starting simulation")
    for i in 1:n_mcs
        if i%100 == 0
            println("MCS = $i")
        end
        mcs(graph, N, p, q, states)
    end
end

function gen_states(N::Int, c::Float64)::Array
    # function that generates initial states
    states = fill(-1, N)
    for i in sample(1:N, Int(c*N), replace=false)
        states[i] = 1
    end
    states
end

function calc_con(states::Array, N::Int)
    return (1/(2*N))*sum(states) + 0.5
end

function save_results(data, q, c)
    println("Saving results")
    open("results_$(q)_$(c).txt", "w") do f
        write(f, string(data))
    end
end

println("Generating graph")
G = erdos_renyi(N, avg_d/(N-1))

function main(graph)
    results = zeros(60)
    i = 0
    # for p in range(0,stop=0.4, length=60)   # for q = 2
    for p in range(0,stop=0.11, length=60)    # for q = 8
        i += 1
        println("Strarting simulation for i=$i, c=$c, q=$q")
        states = gen_states(N, c)
        simulate(
            graph,
            N,
            p,
            q,
            MCS_STEPS,
            states
        )
        results[i] = calc_con(states, N)
        println("Result $(results[i])")
    end
    save_results(results, q, c)
end

main(G)

# @async main(G, 1.0, 2)
# @async main(G, 0.5, 2)
# @async main(G, 1.0, 8)
# @async main(G, 0.5, 8)
